modules/core/src/main/scala/org/tesselation/domain/cluster/programs/Joining.scala

```diff
@@ -19,58 +21,92 @@ import org.tesselation.keytool.security.Signing.verifySignature
 import org.tesselation.keytool.security.{SecurityProvider, hex2bytes}
 import org.tesselation.kryo.KryoSerializer
 import org.tesselation.schema.cluster._
-import org.tesselation.schema.peer.{JoinRequest, Peer, SignRequest}
+import org.tesselation.schema.peer._
 
+import fs2.Pipe
+import fs2.concurrent.Topic
 import org.typelevel.log4cats.slf4j.Slf4jLogger
 
-final case class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerializer](
+/**
+  * Join process:
+  * 1. Node state allows to join
+  * 2. Register peer
+  * 3. Fetch data
+  * 4. Joining height ???
+  *
+  * Register peer:
+  * 1. Create session token
+  * 2. Check whitelisting
+  * 3. Get peer registration request, validate and save it locally
+  * 3. Send own registration request to the peer
+  * 4. Check registration two-way
+  */
+
+object Joining {
+
+  def make[F[_]: Async: GenUUID: SecurityProvider: KryoSerializer](
+    nodeStorage: NodeStorage[F],
+    clusterStorage: ClusterStorage[F],
+    p2pClient: P2PClient[F],
+    cluster: Cluster[F],
+    session: Session[F],
+    peerDiscovery: PeerDiscovery[F]
+  ): F[Joining[F]] =
+    Topic
+      .apply[F, P2PContext]
+      .flatMap(make(_, nodeStorage, clusterStorage, p2pClient, cluster, session, peerDiscovery))
+
+  def make[F[_]: Async: GenUUID: SecurityProvider: KryoSerializer](
+    joiningTopic: Topic[F, P2PContext],
+    nodeStorage: NodeStorage[F],
+    clusterStorage: ClusterStorage[F],
+    p2pClient: P2PClient[F],
+    cluster: Cluster[F],
+    session: Session[F],
+    peerDiscovery: PeerDiscovery[F]
+  ): F[Joining[F]] = {
+    val joining = new Joining(
+      nodeStorage,
+      clusterStorage,
+      p2pClient,
+      cluster,
+      session,
+      joiningTopic
+    ) {}
+
+    def join: Pipe[F, P2PContext, Unit] =
+      in =>
+        in.evalMap { peer =>
+          joining.twoWayHandshake(peer) >>
+            peerDiscovery
+              .discoverFrom(peer)
+              .flatMap { _.toList.traverse(joiningTopic.publish1(_).void) }
+              .void
+        }
+
+    val process = joiningTopic.subscribe(1).through(join).compile.drain
```

Comment:
I wonder if subscribe(1) is a way to go, I guess the topic will get blocked (so no new peers will be enqueued) until this part will process this single element. Not sure if I'm understanding the Topic/Subscriber docs correctly but possibly the flow could get stuck? Also do we need to use the Topic -> Subscriber flow? From reading the fs2 docs some time ago I got a feeling that it's there to allow integration with reactive stream based solutions like Kafka for example, but pure fs2 encourages to write implementations without those - just a feeling, not really sure if it can be done differently.